NOnion/Services/TorServiceHost.fs

```diff
@@ -66,6 +71,85 @@ type TorServiceHost (directory: TorDirectory) =
         }
 
     member private self.RelayIntroduceCallback (introduce: RelayIntroduce) =
+        let rec tryConnectingToRendezvous
+            tryNumber
+            maxRetryCount
+            rendEndpoint
+            rendFingerPrint
+            onionKey
+            cookie
+            clientPubKey
+            introAuthPubKey
+            introEncPrivKey
+            introEncPubKey
+            =
+            async {
+                try
+                    let! endPoint, randomNodeDetails = directory.GetRouter false
+
+                    let! guard = TorGuard.NewClient endPoint
+
+                    let rendCircuit =
+                        TorCircuit (guard, self.IncomingServiceStreamCallback)
+
+                    do! rendCircuit.Create randomNodeDetails |> Async.Ignore
+
+                    do!
+                        rendCircuit.Extend (
+                            CircuitNodeDetail.Create (
+                                rendEndpoint,
+                                onionKey,
+                                rendFingerPrint
+                            )
+                        )
+                        |> Async.Ignore
+
+                    do!
+                        rendCircuit.Rendezvous
+                            cookie
+                            (X25519PublicKeyParameters (clientPubKey, 0))
+                            introAuthPubKey
+                            introEncPrivKey
+                            introEncPubKey
+                with
+                | :? TimeoutException
+                | :? CircuitDestroyedException
+                | :? CircuitTruncatedException
+                | :? GuardConnectionFailedException
```

Comment:
can you make most of these exceptions inherit from a base NOnionException class please?

Reply:
done